File: cpp/s0001_two_sum.cpp
#include <iostream>
#include <vector>
#include <cassert>

using namespace std;


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[i] + nums[j] == target) {
                    return {i, j};
                }
            }
        }
        return {};
    }
};

void runTest(const vector<int>& nums, int target, const vector<int>& expected, const string& testName) {
    Solution solution;
    vector<int> result = solution.twoSum(const_cast<vector<int>&>(nums), target);
    assert(result == expected && (cout << testName << " passed!" << endl));
}

int main() {
    runTest({2, 7, 11, 15}, 9, {0, 1}, "Test Case 1");
    runTest({3,2,4}, 6, {1, 2}, "Test Case 2");
    runTest({3,3}, 6, {0, 1}, "Test Case 3");

    std::cout << "All test cases passed!" << std::endl;

    return 0;
}

File: rust/src/s0001_two_sum.rs
pub struct Solution {}

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        for (index, num) in nums.iter().enumerate() {
            if let Some(sub_index) = nums.iter().position(|&x| x == target-num) {
                if index != sub_index {
                    return vec![index as i32, sub_index as i32]

                }
            }
        }
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(vec![0, 1], Solution::two_sum(vec![2, 7, 11, 15], 9));
        assert_eq!(vec![1, 2], Solution::two_sum(vec![3, 2, 4], 6));
    }
}
